src/shared/sleep-config.c

```diff
@@ -141,6 +144,123 @@ int read_battery_capacity_percentage(void) {
+        if (r == -ENOENT)
+               log_debug_errno(r, "/sys/class/power_supply/BAT0/model_name is unavailable: %m");
+        else if (r < 0)
+               log_debug_errno(r, "Failed to read /sys/class/power_supply/BAT0/model_name: %m");
+
+        r = read_one_line_file("/sys/class/power_supply/BAT0/serial_number", &bat_serial_number);
+        if (r == -ENOENT)
+               log_debug_errno(r, "/sys/class/power_supply/BAT0/serial_number is unavailable: %m");
+        else if (r < 0)
+               log_debug_errno(r, "Failed to read /sys/class/power_supply/BAT0/serial_number: %m");
+
+        r = sd_id128_get_machine(&machine_id);
+        if (r < 0)
+               log_debug_errno(r, "Failed to get machine ID: %m");
+
+        r = id128_get_product(&product_id);
+        if (r < 0)
+                log_debug_errno(r, "Failed to get product ID: %m");
+
+        sha256_init_ctx(&hash);
+        sha256_process_bytes(bat_manufacturer, sizeof(bat_manufacturer), &hash);
+        sha256_finish_ctx(&hash, ret);
+}
+
+static int battery_discharge_rate_in_range(int battery_discharge_rate) {
+        if (battery_discharge_rate <= 0 || battery_discharge_rate >= 200)
+               return 0;
+               /* battery discharge rate is not in a valid range */
+        log_debug("Battery discharge rate is in valid range");
+        return 1;
+}
+
+int get_battery_discharge_rate(void) {
+        _cleanup_free_ char *hash_id_discharge_rate = NULL;
+        const char *p, *filepath;
+        uint8_t *current_hash_id = NULL;
+        char *stored_hash_id = NULL, *stored_discharge_rate = NULL, *h = NULL;
+        int discharge_rate, r;
+
+        filepath = "/var/lib/systemd/sleep/battery_discharge_percentage_rate_per_hour";
+
+        r = read_one_line_file(filepath, &hash_id_discharge_rate);
+        log_debug("Stored hash_id and discharge_rate is %s", hash_id_discharge_rate);
+        if (r < 0)
+               return log_debug_errno(r, "Failed to read discharge rate from %s: %m", filepath);
+
+        get_system_battery_identifier_hash(current_hash_id);
+        log_debug("system_battery_identifier hash is %s", current_hash_id);
+        if (!current_hash_id)
+               return log_debug_errno(r, "Failed to generate system battery identifier hash: %m");
+
+        /* extract the string in file, split it, validate the hash and then return discharge rate */
+
+        p = hash_id_discharge_rate;
+        r = extract_many_words(&p, " ", 0, &stored_hash_id, &stored_discharge_rate, NULL);
+        log_debug("stored hash_id is %s and stored discharge_rate is %s", stored_hash_id, stored_discharge_rate);
+        if (r < 0)
+               return log_debug_errno(r, "Failed to parse hash_id and discharge_rate read from %s location: %m", filepath);
+
+        h = hexmem(current_hash_id, sizeof(current_hash_id));
+        if (!first_word(hash_id_discharge_rate, h))
```

Comment:
no need for first_word here since the first word was already extracted, just compare directly `streq(stored_hash_id, hexmem(...))`